### ISSUE TITLE
[Linux/arm32] GCStress=C Illegal instruction failure

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#21582]
Seen in several runs, not on the same test.

Examples:

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_gcstress0xc_tst/40/consoleFull
```
14:22:05 
      Loader_classloader._TypeGeneratorTests_TypeGeneratorTest716_Generated716_Generated716_._TypeGeneratorTests_TypeGeneratorTest716_Generated716_Generated716_sh [FAIL]
14:22:05         /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_gcstress0xc_tst/bin/tests/Linux.arm.Checked/Loader/classloader/TypeGeneratorTests/TypeGeneratorTest716/Generated716/Generated716.sh: line 244: 26244 Illegal instruction     $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
```

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_gcstress0xc_tst/39/consoleFull
```
11:32:05 
      baseservices_threading._paramthreadstart_ThreadStartOperations_2_ThreadStartOperations_2_._paramthreadstart_ThreadStartOperations_2_ThreadStartOperations_2_sh [FAIL]
11:32:05         ../ThreadStartOperations/ThreadStartOperations.sh: 28: ../ThreadStartOperations/ThreadStartOperations.sh: source: not found
11:32:05         ../ThreadStartOperations/ThreadStartOperations.sh: 167: ../ThreadStartOperations/ThreadStartOperations.sh: shopt: not found
11:32:05         Illegal instruction
11:32:05         /n/nReturn code:      1
```

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_gcstress0xc_tst/38/consoleFull
```
14:48:57       Loader_classloader._TypeGeneratorTests_TypeGeneratorTest1113_Generated1113_Generated1113_._TypeGeneratorTests_TypeGeneratorTest1113_Generated1113_Generated1113_sh [FAIL]
14:48:57         /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_gcstress0xc_tst/bin/tests/Linux.arm.Checked/Loader/classloader/TypeGeneratorTests/TypeGeneratorTest1113/Generated1113/Generated1113.sh: line 244: 31774 Illegal instruction     $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
```

**Comment by BruceForstall** [comment]
@janvorli

**Comment by janvorli** [comment]
I've tried to repro the issue with the Loader/classloader/TypeGeneratorTests/TypeGeneratorTest1038/Generated1038 on my local Odroid XU4 device. I ran 1000 iterations of the test with GC stress C in a loop and didn't hit any issues. So it seems the issue is gone or extremely rare.

**Comment by AndyAyersMS** [comment]
Has not failed the past 4 rolling runs either (going back to Jan 25):

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_gcstress0xc_tst/48/testReport/Loader_classloader/_TypeGeneratorTests_TypeGeneratorTest1038_Generated1038_Generated1038_/history/

I've had some luck reproing intermittent stress failures by loading down the test machine -- say concurrently running `stress -m X` where you pick an X so that CPU is driven to 100%.

**Comment by janvorli** [comment]
@AndyAyersMS I've tried to run with the stress and also tried to run 6 instances of the test in parallel. In both cases, over 300 iterations have passed just fine. I am going to leave it running overnight and see.

**Comment by janvorli** [comment]
I was finally able to repro it after 2900 iterations of the test when running 
8 instances of the test in parallel overnight. The SIGILL seems to reproduce much faster in the `JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b59554/b59554/b59554.exe` and `/baseservices/exceptions/generics/nested-try-catch08/nested-try-catch08.exe` reported as failing in the #21373 and in the recent test runs. I have just tried to run them in parallel with the test mentioned above and got a repro in about 50 iterations several times in a row.

**Comment by janvorli** [comment]
@AndyAyersMS I have spent more time debugging the issue. Since I am unable to repro it when running under a debugger, I've instrumented various places in PAL to see why the illegal instruction is not handled by the runtime during GC stress. And the answer I found is surprising. The issue is that the SIGILL handler is invoked after a call to PALCommonCleanup (which is being called during the process exit). The PALCommonCleanup sets init_count to 0, which causes the PALIsInitialized() to return false. The sigill_handler calls PALIsInitialized() and if it returns false, it doesn't try to handle the illegal instruction in any way and calls the previous handler registered for SIGILL. Which terminates the app with "Illegal instruction".

Now I am trying to figure out how is it possible that our managed code instrumented with the illegal instructions for the GC stress is still running after the PALCommonCleanup was called when the test b59554 doesn't have any secondary managed threads.

**Comment by AndyAyersMS** [comment]
Most every test these days has a secondary managed thread because of EventPipe (see for instance #21461 or #21145).

**Comment by janvorli** [comment]
@AndyAyersMS you are right. I've finally hit the issue under lldb. I can confirm that there is a secondary managed code thread running (most likely the EventPipe one) and the primary thread is already gone. The GC stress instrumentation of code on this secondary thread triggers the SIGILL signal, but since the PALCommonCleanup was already done, the PALIsInitialized returns false and so we don''t try to handle it.

The reason why this issue happens so rarely is most likely the fact that it depends on the timing of the shutdown of the process threads. To hit the issue, the SIGILL signal has to arrive in the very small window between the main thread clearing the init_count and shutdown of the whole process.

I believe that we should not mark PAL as uninitialized at shutdown of the main thread since we keep code running in secondary threads after that point and that code still can be using PAL.

**Comment by janvorli** [comment]
I've tried to remove setting init_count to zero in PALCommonCleanup and couldn't hit the issue anymore. Running over the weekend, about 4000 iterations passed. I am going to fix that, but it is going to be slightly more involved to be a clean fix.
